### PR TITLE
Model/add backtester

### PR DESCRIPTION
**Pull request by zhang-xinyu**
This PR addresses issue #<insert number>
Add:
- `backtest.py`: calculate historical prediction errors
- `timeseries_metrics.py`: calculate errors between prediction and observation timeseries

### Please Check
- [ ] Are the [JSON schemas](https://github.com/covid-projections/covid-data-model/tree/master/api/schemas) updated?
- [ ] Are the [api docs](https://github.com/covid-projections/covid-data-model/blob/master/api/README.V1.md) updated?
- [ ] Are tests passing?

### CODE REVIEW

File: pyseir/backtest/backtest.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+            load_data.load_hospitalization_data_by_state(state_obj.abbr, t0=ref_date)
+        observations['times'] = np.array(observations['times'])
+
+    if hospitalization_data_type is HospitalizationDataType.CUMULATIVE_HOSPITALIZATIONS:
+        observations['current_hosp'] = np.full(observations['times'].shape[0], np.nan)
+        observations['current_hosp'][hospital_times - observations['times'].min()] = np.diff(hospitalizations)
+    elif hospitalization_data_type is HospitalizationDataType.CURRENT_HOSPITALIZATIONS:
+        observations['current_hosp'] = np.full(observations['times'].shape[0], np.nan)
+        observations['current_hosp'][hospital_times - observations['times'].min()] = hospitalizations
+    # create list of t_lists to run model fitter
```
Comment (erccarls) [inline pyseir/backtest/backtest.py]: Comment is confusing here.

File: pyseir/backtest/backtest.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+from pyseir import load_data
+from pyseir.load_data import HospitalizationDataType
+from pyseir.inference.model_fitter import ModelFitter
+from pyseir.utils import get_run_artifact_path, RunArtifact
+from pyseir.backtest.timeseries_metrics import TimeSeriesMetrics, error_type_to_meaning
+
+
+REF_DATE = datetime(year=2020, month=1, day=1)
+
+def load_observations(fips, ref_date=datetime(year=2020, month=1, day=1)):
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:21]: This is file is probably better represented as a class, similar to how we refactored the model fitter. Else we are passing around a bunch of state information needlessly, and parameter config is spread out.

File: pyseir/backtest/backtest.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        state_obj = us.states.lookup(fips)
+        observations['times'], observations['new_cases'], observations['new_deaths'] = \
+            load_data.load_new_case_data_by_state(state_obj.name, ref_date)
+        hospital_times, hospitalizations, hospitalization_data_type = \
+            load_data.load_hospitalization_data_by_state(state_obj.abbr, t0=ref_date)
+        observations['times'] = np.array(observations['times'])
+
+    if hospitalization_data_type is HospitalizationDataType.CUMULATIVE_HOSPITALIZATIONS:
+        observations['current_hosp'] = np.full(observations['times'].shape[0], np.nan)
+        observations['current_hosp'][hospital_times - observations['times'].min()] = np.diff(hospitalizations)
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:60]: Current hospitalizations != new hospitalizations here.   What I have been doing is to instead try to reconstruct new hospitalizations from current by diffing, and then adding back in the expected discharge outflow. You can see an example in infer_rt. 

Also, we might as well move this method into load_data and I can migrate some of the other classes on to this so we don't have to repeat it all over.

File: pyseir/backtest/backtest.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+                                         mf.t_list + mf.fit_results['t0'],
+                                         mf.mle_model.results['total_deaths_per_day'])
+
+    if mf.hospitalization_data_type is not None:
+        if mf.hospitalization_data_type is HospitalizationDataType.CUMULATIVE_HOSPITALIZATIONS:
+            predicted_hosp = (mf.mle_model.results['HGen_cumulative'] + mf.mle_model.results['HICU_cumulative'])
+            predicted_hosp = np.diff(predicted_hosp)
+        elif mf.hospitalization_data_type is HospitalizationDataType.CURRENT_HOSPITALIZATIONS:
+            predicted_hosp = mf.mle_model.results['HGen'] + mf.mle_model.results['HICU']
+        prediction['current_hosp'] = np.interp(observations['times'],
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:139]: Same comment as above re: current vs new admissions.

File: pyseir/backtest/backtest.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+    prediction = pd.DataFrame(prediction, index=observations.index).dropna(axis=1, how='all')
+    
+    return prediction
+
+
+def run_backtest(fips,
+                 observations,
+                 rolling_window_size=1,
+                 prediction_window_size=7,
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:153]: Seems we should extend this to 2 weeks.

File: pyseir/backtest/backtest.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+        prediction['current_hosp'] = None
+
+    prediction = pd.DataFrame(prediction, index=observations.index).dropna(axis=1, how='all')
+    
+    return prediction
+
+
+def run_backtest(fips,
+                 observations,
+                 rolling_window_size=1,
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:152]: Make this a bit larger?

File: pyseir/backtest/backtest.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    depending on type of errors.
+
+    Parameters
+    ----------
+    fips: str
+        FIPS code of the state or county to run backtest.
+    rolling_window_size: int
+        Width of time window to aggregate (average) the observation and
+        prediction to calculate prediction error.
+    prediction_window_size: int
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:174]: Did you get the chance to use the Gaussian kernel?

File: pyseir/backtest/backtest.py [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+        - new_deaths: float, predicted new deaths
+        - curret_hosp: float, predicted hospitalizations
+        - observation_end_date: Datetime, last day before blinded observations.
+        - observation_days_blinded: int, number of days that observations
+                                    are blinded.
+    """
+    tsm = TimeSeriesMetrics()
+    backtest_results = list()
+    historical_predictions = list()
+    for d in range(1, max_observation_days_blinded + 1):
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:212]: Not sure how long this takes, but could probably add a `step_size` parameter here to speed things up, especially if we are using more smoothing.

File: pyseir/backtest/backtest.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+        observation_end_date = ref_date + timedelta(int(observations['times'].values[-d]))
+        # record predictions
+        prediction = run_model_fitter_for_backtest(fips=fips, observations=observations,
+                                                   observation_days_blinded=d, n_retries=n_retries)
+        prediction['observation_end_date'] = observation_end_date
+        prediction['observation_days_blinded'] = d
+
+        # record back test errors
+        backtest_record = dict()
```
Comment (erccarls) [inline pyseir/backtest/backtest.py:221]: Could use `defaultdict(list)` here

File: pyseir/backtest/backtest.py [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+from multiprocessing import Pool
+from functools import partial
+from pyseir import load_data
+from pyseir.load_data import HospitalizationDataType
+from pyseir.inference.model_fitter import ModelFitter
+from pyseir.utils import get_run_artifact_path, RunArtifact
+from pyseir.backtest.timeseries_metrics import TimeSeriesMetrics, error_type_to_meaning
+
+
+REF_DATE = datetime(year=2020, month=1, day=1)
```
[the remark this answers is not in the document]
Comment (zhang-xinyu) [reply pyseir/backtest/backtest.py:19]: Sure, definitely better keep it consistent. 👍

File: pyseir/backtest/backtest.py (hunk 1 above)
Comment (zhang-xinyu) [reply pyseir/backtest/backtest.py]: Ah right, just removed, thanks.

File: pyseir/backtest/backtest.py (hunk 7 above)
Comment (zhang-xinyu) [reply pyseir/backtest/backtest.py:173]: Yes, just added, thanks!

File: pyseir/backtest/backtest.py (hunk 3 above)
Comment (zhang-xinyu) [reply pyseir/backtest/backtest.py:60]: Ah, got it, let me see if I can add this to load_data.

**Pull request merged**